File: code_practice/LongestCommonSubstring/lcs.py
def solution(s1, s2) -> int:
    len_lcs = 0
    for i in range(len(s1)):
        for j in range(len(s2)):
            ptr = 0
            while (i+ptr < len(s1)) and (j+ptr < len(s2)) and s1[i+ptr] == s2[j+ptr]:
                if ptr > len_lcs:
                    len_lcs = ptr
                ptr += 1
            if ptr > len_lcs:
                len_lcs = ptr
    return len_lcs


def df_solution(s1, s2):
    """
    字符串型
    A[i-1] B[j-1]
    if A[i]B[j] ? +1 : +0
    acde abcde
    """
    dp = [[0 for _ in range(len(s2))] for _ in range(len(s1))]
    for i in range(len(s1)):
        for j in range(len(s2)):
            if s1[i] == s2[j]:
                dp[i][j] += dp[i-1][j-1] + 1
    return max(map(lambda x: max(x), dp))



if __name__ == "__main__":
    print(solution('abcd', 'adbc'))
    print(df_solution('abcd', 'adbcdced'))

File: code_practice/longest_common_substring/lcs.scala
def solution(s1: String, s2: String) : Int = {
    var lcs = 0
    for (i <- 0 to s1.size){
        for (j <- 0 to s2.size){
            var pts = 0
            while (i+pts < s1.size && j+pts < s2.size && s1(i+pts) == s2(j+pts)) {
                pts += 1
            }
            if (pts > lcs) lcs=pts
        }
    }
    lcs
}

println(solution("abcde", "abcdef"))

def df_solution(s1: String, s2: String) : Int = {
    val dp = Array.ofDim[Int](s2.size, s1.size)
    for (i <- 0 to s1.size) {
        for (j <- 0 to s2.size){
            if (s1(i) == s2(j)) dp(i)(j) = dp(i-1)(j-1) + 1
        }
    }
    dp.map(_.reduce((x,y)=>x max y)).reduce((x,y)=>x max y)
}

println(solution("abcde", "abdabcdef"))